pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -204,40 +186,41 @@ public void run() {
                 // process the message
                 Object input;
                 try {
-                    input = msg.getInputSerDe().deserialize(msg.getActualMessage().getData());
+                    input = currentMessage.getInputSerDe().deserialize(currentMessage.getActualMessage().getData());
                 } catch (Exception ex) {
-                    stats.incrementDeserializationExceptions(msg.getTopicName());
-                    continue;
+                    stats.incrementDeserializationExceptions(currentMessage.getTopicName());
+                    throw ex;
                 }
                 long processAt = System.currentTimeMillis();
                 stats.incrementProcessed(processAt);
                 addLogTopicHandler();
-                result = javaInstance.handleMessage(
-                        msg.getActualMessage().getMessageId(),
-                        msg.getTopicName(),
+                JavaExecutionResult result = javaInstance.handleMessage(
+                        currentMessage.getActualMessage().getMessageId(),
+                        currentMessage.getTopicName(),
                         input);
                 removeLogTopicHandler();
 
                 long doneProcessing = System.currentTimeMillis();
                 log.debug("Got result: {}", result.getResult());
 
                 if (null != stateContext) {
-                    stateContext.flush()
-                            .thenRun(() -> processResult(msg, result, processAt, doneProcessing))
-                            .exceptionally(cause -> {
-                                // log the messages, since we DONT ack, pulsar consumer will re-deliver the messages.
-                                log.error("Failed to flush the state updates of message {}", msg, cause);
-                                return null;
-                            });
-                } else {
-                    processResult(msg, result, processAt, doneProcessing);
+                    CompletableFuture completableFuture = stateContext.flush();
```

Comment:
we are sync here as opposed to async before?

Reply:
since we are going to remove `flush` anyway, so it is okay to convert it back to sync for now. it will make the abstraction much easier.